ReturnAndSet won't compile with different argument types

Hi,

given the following snippet:
`
    class X {
    public:
        virtual int f(std::string&, std::vector<std::string>&) = 0;
    };
    Mock<X> mockX;
    std::vector<std::string> d{"S"};
    When(Method(mockX, f)).ReturnAndSet(20, std::placeholders::_2 <= d);
`

Results the following compiler error:
`
esser-thirdparty/FakeIt/single_header/boost/fakeit.hpp:7730:58: error: no viable overloaded '=' [clang-diagnostic-error]
                    GetArg(std::forward<current_arg>(p)) = std::get<check_index>(arg_vals).value;
                                                         ^
FakeIt/single_header/boost/fakeit.hpp:7746:54: note: in instantiation of function template specialization 'fakeit::helper::ArgLocator<1, 0>::AssignArg<std::basic_string<char> &, std::vector<std::basic_string<char>> &, 2>' requested here
            ArgLocator<arg_index, sizeof...(N) - 1>::AssignArg(std::forward<current_arg>(p), arg_vals);
                                                     ^
FakeIt/single_header/boost/fakeit.hpp:7506:29: note: in instantiation of function template specialization 'fakeit::helper::Assign<1, std::basic_string<char> &, std::vector<std::basic_string<char>> &, 2, std::vector<std::basic_string<char>> &>' requested here
                    helper::Assign<1>(vals_tuple, std::forward<arglist>(args)...);
                            ^
FakeIt/single_header/boost/fakeit.hpp:7498:19: note: in instantiation of member function 'fakeit::MethodStubbingProgress<int, std::basic_string<char> &, std::vector<std::basic_string<char>> &>::GetAssigner(int &&, helper::ArgValue<vector<basic_string<char>> &, 2>)::Lambda::operator()' requested here
            class Lambda {
                  ^
FakeIt/single_header/boost/fakeit.hpp:7427:23: note: in instantiation of function template specialization 'fakeit::MethodStubbingProgress<int, std::basic_string<char> &, std::vector<std::basic_string<char>> &>::GetAssigner<std::vector<std::basic_string<char>> &, 2>' requested here
            return Do(GetAssigner(std::forward<R>(r),
                      ^
.cpp:46:28: note: in instantiation of function template specialization 'fakeit::MethodStubbingProgress<int, std::basic_string<char> &, std::vector<std::basic_string<char>> &>::ReturnAndSet<fakeit::helper::ArgValue<std::vector<std::basic_string<char>> &, 2>>' requested here
    When(Method(mockX, f)).ReturnAndSet(20, std::placeholders::_2 <= d);
                           ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/basic_string.h:665:7: note: candidate function not viable: no known conversion from 'std::vector<std::basic_string<char>>' to 'const std::basic_string<char>' for 1st argument
      operator=(const basic_string& __str)
      ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/basic_string.h:675:7: note: candidate function not viable: no known conversion from 'std::vector<std::basic_string<char>>' to 'const char *' for 1st argument
      operator=(const _CharT* __s)
      ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/basic_string.h:686:7: note: candidate function not viable: no known conversion from 'std::vector<std::basic_string<char>>' to 'char' for 1st argument
      operator=(_CharT __c)
      ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/basic_string.h:703:7: note: candidate function not viable: no known conversion from 'std::vector<std::basic_string<char>>' to 'std::basic_string<char>' for 1st argument
      operator=(basic_string&& __str)
      ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/basic_string.h:771:7: note: candidate function not viable: no known conversion from 'std::vector<std::basic_string<char>>' to 'initializer_list<char>' for 1st argument
      operator=(initializer_list<_CharT> __l)
      ^
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/basic_string.h:785:8: note: candidate template ignored: requirement '__and_<std::is_convertible<const std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char>>, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char>>>> &, std::basic_string_view<char, std::char_traits<char>>>, std::__not_<std::is_convertible<const std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char>>, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char>>>> *, const std::basic_string<char, std::char_traits<char>, std::allocator<char>> *>>, std::__not_<std::is_convertible<const std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char>>, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char>>>> &, const char *>>>::value' was not satisfied [with _Tp = std::vector<std::basic_string<char>>]
       operator=(const _Tp& __svt)
       ^
4 warnings and 4 errors generated.
`
But if I change the interface method signature, to match in type of arguments, everything compiles well:
`class X {
public:
    virtual int f(std::vector<std::string>&, std::vector<std::string>&) = 0;
};
`

About env:
- CMake: 3.20
- g++ / gcc 11
- c++20 standard

Isn't it possible to use ReturnAndSet if method has different types of arguments?

--- a/include/fakeit/StubbingProgress.hpp
+++ b/include/fakeit/StubbingProgress.hpp
@@ -436,6 +436,7 @@ namespace fakeit {
                     ArgLocator<arg_index, check_index - 1>::AssignArg(std::forward<current_arg>(p), arg_vals);
             }
 
+#if __cplusplus < 201703L
         private:
             template<typename T, typename U>
             static
@@ -452,6 +453,7 @@ namespace fakeit {
             {
                 throw std::logic_error("ReturnAndSet(): Invalid value type");
             }
+#endif
 
         };
 

--- a/include/fakeit/StubbingProgress.hpp
+++ b/include/fakeit/StubbingProgress.hpp
@@ -425,11 +425,34 @@ namespace fakeit {
         struct ArgLocator {
             template<typename current_arg, typename ...T, int ...N>
             static void AssignArg(current_arg &&p, std::tuple<ArgValue<T, N>...> arg_vals) {
-                if (std::get<check_index>(arg_vals).pos == arg_index)
+#if __cplusplus >= 201703L
+                if constexpr (std::get<check_index>(arg_vals).pos == arg_index)
                     GetArg(std::forward<current_arg>(p)) = std::get<check_index>(arg_vals).value;
+#else
+                if (std::get<check_index>(arg_vals).pos == arg_index)
+                    Set(std::forward<current_arg>(p), std::get<check_index>(arg_vals).value);
+#endif
                 else if (check_index > 0)
                     ArgLocator<arg_index, check_index - 1>::AssignArg(std::forward<current_arg>(p), arg_vals);
             }
+
+        private:
+            template<typename T, typename U>
+            static
+            typename std::enable_if<std::is_convertible<U, decltype(GetArg(std::declval<T>()))>::value, void>::type
+            Set(T &&p, U &&v)
+            {
+                GetArg(std::forward<T>(p)) = v;
+            }
+
+            template<typename T, typename U>
+            static
+            typename std::enable_if<!std::is_convertible<U, decltype(GetArg(std::declval<T>()))>::value, void>::type
+            Set(T &&, U &&)
+            {
+                throw std::logic_error("ReturnAndSet(): Invalid value type");
+            }
+
         };
 
         template<int arg_index>

--- a/tests/stubbing_tests.cpp
+++ b/tests/stubbing_tests.cpp
@@ -26,6 +26,7 @@ struct BasicStubbing : tpunit::TestFixture {
                     TEST(BasicStubbing::stub_a_function_to_return_a_specified_value_always),
                     TEST(BasicStubbing::stub_a_function_to_set_specified_values_once),
                     TEST(BasicStubbing::stub_a_function_to_set_specified_values_once_form2),
+                    TEST(BasicStubbing::stub_a_function_to_set_specified_value_with_incompatible_params),
                     TEST(BasicStubbing::stub_a_function_to_set_specified_values_always),
                     TEST(BasicStubbing::stub_a_function_to_set_specified_values_always_form2),
                     TEST(BasicStubbing::stub_a_method_to_throw_a_specified_exception_once),//
@@ -167,8 +168,6 @@ struct BasicStubbing : tpunit::TestFixture {
         Mock<SomeInterface> mock;
         When(Method(mock, funcRefArgs)).ReturnAndSet(1, _2 <= 3, _1 <= 2);
         When(Method(mock, procRefArgs)).ReturnAndSet(_1 <= 4, _2 <= 5).ReturnAndSet( _2 <= 6, _1 <= 7).ReturnAndSet(_2 <= 8);
-        std::vector<std::string> v{"str"};
-        When(Method(mock, procIncompatArgs)).ReturnAndSet(_2 <= v);
 
         SomeInterface &i = mock.get();
 
@@ -196,11 +195,28 @@ struct BasicStubbing : tpunit::TestFixture {
             FAIL();
         } catch (fakeit::UnexpectedMethodCallException &) {
         }
+    }
+
+    void stub_a_function_to_set_specified_value_with_incompatible_params() {
+        Mock<SomeInterface> mock;
+        std::vector<std::string> v{"str"};
+        When(Method(mock, procIncompatArgs)).ReturnAndSet(_2 <= v);
+
+        SomeInterface &i = mock.get();
 
         std::string s;
         std::vector<std::string> chk_v;
         i.procIncompatArgs(s, chk_v);
         ASSERT_EQUAL(chk_v, v);
+
+#if __cplusplus < 201703L
+        When(Method(mock, procIncompatArgs)).ReturnAndSet(_1 <= v);
+        try {
+            i.procIncompatArgs(s, chk_v);
+            FAIL();
+        } catch (std::logic_error&) {
+        }
+#endif
     }
 
     void stub_a_function_to_set_specified_values_always() {

--- a/tests/stubbing_tests.cpp
+++ b/tests/stubbing_tests.cpp
@@ -63,6 +63,7 @@ struct BasicStubbing : tpunit::TestFixture {
 
         virtual void proc(int) = 0;
         virtual void procRefArgs(int*, int&) = 0;
+        virtual void procIncompatArgs(std::string&, std::vector<std::string>&) = 0;
     };
 
     void calling_an_unstubbed_method_should_raise_UnmockedMethodCallException() {
@@ -166,6 +167,8 @@ struct BasicStubbing : tpunit::TestFixture {
         Mock<SomeInterface> mock;
         When(Method(mock, funcRefArgs)).ReturnAndSet(1, _2 <= 3, _1 <= 2);
         When(Method(mock, procRefArgs)).ReturnAndSet(_1 <= 4, _2 <= 5).ReturnAndSet( _2 <= 6, _1 <= 7).ReturnAndSet(_2 <= 8);
+        std::vector<std::string> v{"str"};
+        When(Method(mock, procIncompatArgs)).ReturnAndSet(_2 <= v);
 
         SomeInterface &i = mock.get();
 
@@ -193,6 +196,11 @@ struct BasicStubbing : tpunit::TestFixture {
             FAIL();
         } catch (fakeit::UnexpectedMethodCallException &) {
         }
+
+        std::string s;
+        std::vector<std::string> chk_v;
+        i.procIncompatArgs(s, chk_v);
+        ASSERT_EQUAL(chk_v, v);
     }
 
     void stub_a_function_to_set_specified_values_always() {

